File: app/src/main/java/org/ole/planet/myplanet/ui/feedback/AdapterFeedback.java
package org.ole.planet.myplanet.ui.feedback;

import android.content.Context;
import android.content.Intent;
import android.view.LayoutInflater;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import org.ole.planet.myplanet.R;
import org.ole.planet.myplanet.databinding.RowFeedbackBinding;
import org.ole.planet.myplanet.model.RealmFeedback;
import org.ole.planet.myplanet.utilities.TimeUtils;

import java.util.List;

import io.realm.RealmResults;

public class AdapterFeedback extends RecyclerView.Adapter<AdapterFeedback.ViewHolderFeedback> {
    private RowFeedbackBinding rowFeedbackBinding;
    private Context context;
    private List<RealmFeedback> list;

    public AdapterFeedback(Context context, List<RealmFeedback> list) {
        this.context = context;
        this.list = list;
    }

    public void updateData(RealmResults<RealmFeedback> newData) {
        list = newData;
        notifyDataSetChanged();
    }

    @NonNull
    @Override
    public ViewHolderFeedback onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        rowFeedbackBinding = RowFeedbackBinding.inflate(LayoutInflater.from(context), parent, false);
        return new ViewHolderFeedback(rowFeedbackBinding);
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolderFeedback holder, int position) {
        rowFeedbackBinding.tvTitle.setText(list.get(position).title);
        rowFeedbackBinding.tvType.setText(list.get(position).type);
        rowFeedbackBinding.tvPriority.setText(list.get(position).priority);
        rowFeedbackBinding.tvStatus.setText(list.get(position).status);
        if ("yes".equalsIgnoreCase(list.get(position).priority))
            rowFeedbackBinding.tvPriority.setBackground(context.getResources().getDrawable(R.drawable.bg_primary));
        else
            rowFeedbackBinding.tvPriority.setBackground(context.getResources().getDrawable(R.drawable.bg_grey));
        rowFeedbackBinding.tvStatus.setBackground(context.getResources().getDrawable("open".equalsIgnoreCase(list.get(position).status) ? R.drawable.bg_primary : R.drawable.bg_grey));
        rowFeedbackBinding.tvOpenDate.setText(TimeUtils.getFormatedDate(Long.parseLong(list.get(position).openTime)));
        rowFeedbackBinding.getRoot().setOnClickListener(v -> context.startActivity(new Intent(context, FeedbackDetailActivity.class).putExtra("id", list.get(position).id)));
    }

    @Override
    public int getItemCount() {
        return list.size();
    }

    public static class ViewHolderFeedback extends RecyclerView.ViewHolder {
        RowFeedbackBinding rowFeedbackBinding;

        public ViewHolderFeedback(RowFeedbackBinding rowFeedbackBinding) {
            super(rowFeedbackBinding.getRoot());
            this.rowFeedbackBinding = rowFeedbackBinding;
        }
    }
}

File: app/src/main/java/org/ole/planet/myplanet/ui/feedback/AdapterFeedback.kt
package org.ole.planet.myplanet.ui.feedback

import android.content.Intent
import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.core.content.res.ResourcesCompat
import androidx.recyclerview.widget.ListAdapter
import androidx.recyclerview.widget.RecyclerView
import org.ole.planet.myplanet.R
import org.ole.planet.myplanet.databinding.RowFeedbackBinding
import org.ole.planet.myplanet.model.RealmFeedback
import org.ole.planet.myplanet.ui.feedback.AdapterFeedback.ViewHolderFeedback
import org.ole.planet.myplanet.utilities.DiffUtils
import org.ole.planet.myplanet.utilities.TimeUtils.getFormattedDate

class AdapterFeedback :
    ListAdapter<RealmFeedback, ViewHolderFeedback>(
        DiffUtils.itemCallback(
            { oldItem, newItem ->
                oldItem.id == newItem.id
            },
            { oldItem, newItem ->
                oldItem.title == newItem.title &&
                    oldItem.type == newItem.type &&
                    oldItem.priority == newItem.priority &&
                    oldItem.statusEnum == newItem.statusEnum &&
                    oldItem.openTime == newItem.openTime
            }
        )
    ) {

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolderFeedback {
        val binding = RowFeedbackBinding.inflate(LayoutInflater.from(parent.context), parent, false)
        return ViewHolderFeedback(binding)
    }

    override fun onBindViewHolder(holder: ViewHolderFeedback, position: Int) {
        val feedback = getItem(position)
        val binding = holder.rowFeedbackBinding
        val context = binding.root.context

        binding.tvTitle.text = feedback.title
        binding.tvType.text = feedback.type
        binding.tvPriority.text = feedback.priority
        val status = feedback.statusEnum
        val statusText = status.displayText(context)
        binding.tvStatus.text = statusText
        val contentDescription = "${feedback.title}, ${feedback.type}, " +
                "${context.getString(R.string.status)}: $statusText, ${context.getString(R.string.priority)}: ${feedback.priority}, " +
                "${context.getString(R.string.open_date)}: ${getFormattedDate(feedback.openTime)}"
        binding.feedbackCardView.contentDescription = contentDescription

        if ("yes".equals(feedback.priority, ignoreCase = true)) {
            binding.tvPriority.background =
                ResourcesCompat.getDrawable(context.resources, R.drawable.bg_primary, null)
        } else {
            binding.tvPriority.background =
                ResourcesCompat.getDrawable(context.resources, R.drawable.bg_grey, null)
        }
        binding.tvStatus.background = ResourcesCompat.getDrawable(
            context.resources,
            status.backgroundRes,
            null
        )
        binding.tvOpenDate.text = getFormattedDate(feedback.openTime)
        binding.root.setOnClickListener {
            binding.root.contentDescription = feedback.title
            context.startActivity(
                Intent(context, FeedbackDetailActivity::class.java)
                    .putExtra("id", feedback.id)
            )
        }
    }

    class ViewHolderFeedback(val rowFeedbackBinding: RowFeedbackBinding) :
        RecyclerView.ViewHolder(rowFeedbackBinding.root)
}
